pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1212,6 +1221,33 @@ protected void handleSend(CommandSend send, ByteBuf headersAndPayload) {
 
         startSendOperation(producer, headersAndPayload.readableBytes(), send.getNumMessages());
 
+        final long produceId = producer.getProducerId();
```

Comment:
It's better to handle the transaction message in the topic, not the server cnx.

Reply:
Yes, that's more reasonable.